ledger-core/inc/core/database/DatabaseSessionPool.hpp

```diff
@@ -75,20 +75,48 @@ namespace ledger {
                 soci::transaction tr(sql);
                 Migration<T::CURRENT_VERSION, T>::forward(sql, version);
 
+                // register rollback migration
+                _rollbackMigrations.emplace_back([](soci::session& sql) {
+                    int version = getDatabaseMigrationVersion<T>(sql);
+
+                    soci::transaction tr(sql);
+                    Migration<T::CURRENT_VERSION, T>::backward(sql, version);
+                    tr.commit();
+                });
+
                 tr.commit();
             }
 
-            /// Rollback a migration for a given migration system.
-            template <typename T>
-            void rollbackMigration() {
+            /// Run migration forward for the ledger-core library.
+            ///
+            /// This is a special case because we don’t want to register ourselves for rollback,
+            /// since we will call that function explicitly.
+            template <>
+            void forwardMigration<CoreMigration>() {
                 soci::session sql(getPool());
-                int version = getDatabaseMigrationVersion<T>(sql);
+                int version = getDatabaseMigrationVersion<CoreMigration>(sql);
 
                 soci::transaction tr(sql);
-                Migration<T::CURRENT_VERSION, T>::backward(sql, version);
+                Migration<CoreMigration::CURRENT_VERSION, CoreMigration>::forward(sql, version);
 
-                unsetupMigrations(sql);
+                tr.commit();
+            }
 
+            /// Rollback all migrations for a given migration system.
+            void rollbackMigrations() {
+                soci::session sql(getPool());
+
+                // rollback coins first
+                for (auto rollbackCoinMigration : _rollbackMigrations) {
```

Comment:
You can avoid a copy by taking reference values in the range-based for loop :
```
for (auto& rollbackMigration : _rollbackMigrations)
```
Or use this beautiful line of code :
```
std::for_each(
   rollbackMigrations.cbegin(), 
   rollbackMigrations.cend(), 
   std::bind(
      &decltype(rollbackMigrations)::value_type::operator(),
      std::placeholders::_1,
      std::ref(sql)
   )
);
```
:trollface: